prov/util/src/util_ns.c

```diff
@@ -0,0 +1,500 @@
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/*
+ * A name server is started on each
+ * node as a thread within one of the processes on that node. It maintains
+ * a database that maps "services" to "endpoint names". Other processes
+ * on the same node talk to this name server to update mapping information.
+ *
+ * To resolve a "node:service" pair into an provider internal endpoint name
+ * that can be used as the input of fi_av_insert, a process needs to make
+ * a query to the name server residing on "node".
+ */
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+
+#include <fi_util.h>
+#include <rdma/providers/fi_log.h>
+#include <fi.h>
+
+#include "rbtree.h"
+
+#define OFI_UTIL_NS_DEFAULT_PORT	5678
+
+enum {
+	OFI_UTIL_NS_ADD,
+	OFI_UTIL_NS_DEL,
+	OFI_UTIL_NS_QUERY,
+	OFI_UTIL_NS_ACK,
+};
+
+struct util_ns_cmd {
+	int	op;
+	int	service;
+	int	status;
+	size_t	name_len;
+};
+
+struct util_ns_entry {
+	size_t len;
+	void *buf;
+};
+
+static int util_ns_map_init(struct util_ns *ns)
+{
+	if (!ns->service_comp)
+		return -FI_EINVAL;
+
+	ns->ns_map = rbtNew(ns->service_comp);
+
+	return ns->ns_map ? 0 : -FI_ENOMEM;
+}
+
+static void util_ns_map_fini(struct util_ns *ns)
+{
+	rbtDelete(ns->ns_map);
+}
+
+static int util_ns_map_add(struct util_ns *ns, int service,
+			   void *name_in, size_t name_len)
+{
+	int ret = FI_SUCCESS;
+	struct util_ns_entry *entry = calloc(1, sizeof(struct util_ns_entry));
+	if (!entry)
+		return -FI_ENOMEM;
+
+	if (rbtFind(ns->ns_map, (void *)(uintptr_t)service)) {
+		ret = -FI_EBUSY;
+		goto err;
+	}
+
+	entry->len = name_len;
+	entry->buf = name_in;
+
+	if (rbtInsert(ns->ns_map, (void *)(uintptr_t)service,
+		      (void *)entry)) {
+		ret = -FI_ENOMEM;
+		goto err;
+	}
+
+	return FI_SUCCESS;
+err:
+	free(entry);
+	return ret;
+}
+
+static int util_ns_map_del(struct util_ns *ns, int service,
+			   char *name_in, size_t name_len)
+{
+	RbtIterator it;
+	void *key;
+	struct util_ns_entry *entry;
+
+        it = rbtFind(ns->ns_map, (void *)(uintptr_t)service);
+        if (it) {
+		rbtKeyValue(ns->ns_map, it, &key, (void **)&entry);
+		if (name_len != entry->len ||
+		    memcmp(entry->buf, name_in, name_len))
+			return -FI_ENOENT;
+		free(entry->buf);
+		free(entry);
+		rbtErase(ns->ns_map, it);
+	}
+
+	return FI_SUCCESS;
+}
+
+static int util_ns_map_lookup(struct util_ns *ns, int *service,
+			      void **name_out, size_t *name_len)
+{
+	RbtIterator it;
+	void *key;
+	struct util_ns_entry *entry;
+
+        it = rbtFind(ns->ns_map, (void *)(uintptr_t)(*service));
+	if (!it)
+		return -FI_ENOENT;
+
+	rbtKeyValue(ns->ns_map, it, &key, (void **)&entry);
+	*name_out = entry->buf;
+	*name_len = entry->len;
+
+	if (*service == 0)
```

Comment:
yes, that's what I mean. that compare function is generic enough if 0 for "any service" works for all providers.